[13.0] Add stock_checkout_sync

guewen opened the pull request:
Replaces #862 

Regroup moves at the same location when they reach an operation type.

Example of use case:

* We use several operation types for Picking Transfers (Pick Shelves, Pick Reserve)
* We use several operation types for Packing (Pack, Pack Post)
* The "Checkout Synchronization" flag is activated on the 2 Packing operation types
* The Packing can occur in several areas (Packing/PACK1, Packing/PACK2, Packing/PACK3)

Based on these assumptions, we have these pickings, all for the same order:

* SHELF001, pick 2 goods in the shelves
* RESER001, pick 2 goods in the reserve
* PACK001, pack 3 of the goods (2 from SHELF001, 1 from RESER001)
* PPOST001, pack the remaining good of the reserve

The operator uses the button "Checkout Sync" on a pick transfer,
for instance SHELF001, and select Packing/PACK2 as destination.
We want all the lines that reach PACK001 to be updated with:

* The move lines (except the one that goes to PPOST001) must now reach
  Packing/PACK2
* The same moves' destination must reach Packing/PACK2 as well:
  we constrain the user to move the goods there as we want to move
  all the goods at the same place for the packing.

![Selection_848](https://user-images.githubusercontent.com/417223/82321125-a34bed00-99d4-11ea-84c1-7138749b6ab9.png)

Works best when used with ``stock_available_to_promise_release``:

* When using the mentioned module, we assume that we release moves (which
  creates the whole chain of moves) only when we know that we have the
  quantities in stock (otherwise the module splits the delivery). Any exception
  (which is now rare when you release this ways) that would lead to a backorder
  is handled by canceling the remaining line.
* When the module is not used, then the destination of the backorders may be
  changed at the same time

Compatible with ``stock_dynamic_routing``.

Commit:
Add stock_checkout_sync

Add an assistant to select the same destination location for all the
moves that will reach the same transfer. Generally used on packing
locations.

simahawk commented on stock_checkout_sync/models/stock_move.py:
```diff
@@ -0,0 +1,54 @@
+            for dest_picking in dest_pickings:
+                if dest_picking not in selected_pickings:
+                    selected_pickings[dest_picking] = self.env["stock.move"].browse()
+                selected_pickings[dest_picking] |= moves
```
shall we use ids here instead of recordsets?

guewen replied:
I don't really expect to have thousands of ids to make a difference here. It's used on one order, so a dozen of moves, maybe a hundred for a very large order.

Commit:
Fix tests, picking type is now mandatory on routing

grindtildeath commented on stock_checkout_sync/models/stock_picking.py:
```diff
@@ -0,0 +1,25 @@
+                    break
+
+    def open_checkout_sync_wizard(self, done_picking_types=None):
+        return self.env["stock.move.checkout.sync"]._open(self)
```
Isn't it simpler to rely on `env.context.active_id` in `default_get` of the wizard and return an action here?

guewen replied:
The wizard reopens itself, that's why the action is returned by a method `_open` in the wizard class. Then we need a button on the picking, and we can reuse the `_open` method of the wizard as it exists.

grindtildeath commented on stock_checkout_sync/models/stock_picking.py:
```diff
@@ -0,0 +1,25 @@
+    def _compute_can_sync_to_checkout(self):
+        for picking in self:
+            picking.can_sync_to_checkout = False
+            for dest_move in picking.mapped("move_lines.move_dest_ids"):
+                if dest_move.picking_type_id.checkout_sync:
+                    picking.can_sync_to_checkout = True
+                    break
```
```suggestion
            if any(
                dest_move.picking_type_id.checkout_sync for dest_move in picking.mapped(
                    "move_lines.move_dest_ids"
                )
            ):
                    picking.can_sync_to_checkout = True
```
Not sure if this makes the condition more "readable"

grindtildeath commented on stock_move_common_dest/models/stock_move.py:
```diff
@@ -47,3 +52,16 @@ def _compute_common_dest_move_ids(self):
             else:
                 move.common_dest_move_ids = [(5, 0, 0)]
+
+    def _search_compute_dest_move_ids(self, operator, value):
```
```suggestion
    def _search_common_dest_move_ids(self, operator, value):
```

grindtildeath commented on stock_move_common_dest/models/stock_move.py:
```diff
@@ -10,17 +10,12 @@ class StockMove(models.Model):
     common_dest_move_ids = fields.Many2many(
         "stock.move",
         compute="_compute_common_dest_move_ids",
+        search="_search_compute_dest_move_ids",
```
```suggestion
        search="_search_common_dest_move_ids",
```

grindtildeath commented on stock_checkout_sync/models/stock_move.py:
```diff
@@ -0,0 +1,54 @@
+            lambda l: l.location_dest_id != location and l.state != "done"
+        ).write({"location_dest_id": location.id})
+
+    def _moves_to_sync_checkout(self):
```
```suggestion
    def _select_pickings_to_sync_checkout(self):
```
I find it odd to return picking with the actual function name

guewen replied:
It does return moves, but grouped by pickings in a dictionary.

grindtildeath reviewed (commented):
Looks good overall, just some functions names to fix and questions.

Commit:
Apply suggestions from code review

Co-authored-by: Akim Juillerat <akim.juillerat@camptocamp.com>

simahawk commented:
@guewen had to solve a conflict to make pending merge aggregation work. Fill free to rebase and trash https://github.com/OCA/stock-logistics-warehouse/pull/907/commits/314f5fba683e9147005635b5cf8adb72a4e437d5

Commit:
Add a test ensuring new move lines have correct destination

sebalix commented on stock_checkout_sync/models/stock_move.py:
```diff
@@ -0,0 +1,54 @@
+            lambda m: (
+                m.state == "waiting"
+                or m.state == "assigned"
+                and m in self.move_dest_ids
```
I'm not sure to understand this line, `m` is supposed to be part of `moves_to_update.move_dest_ids`, and `moves_to_update` is already a subset of `self` so it's always True, no?

guewen replied:
You are right, thanks

guewen commented on stock_checkout_sync/models/stock_move.py:
```diff
@@ -0,0 +1,54 @@
+            lambda m: (
+                m.state == "waiting"
+                or m.state == "assigned"
+                and m in self.move_dest_ids
```
```suggestion
```

Commit:
Remove useless check (code review)

m is supposed to be part of moves_to_update.move_dest_ids, and
moves_to_update is already a subset of self so it's always True

Commit:
Synchronize destination location on package level

sebalix commented on stock_dynamic_routing_checkout_sync/__manifest__.py:
```diff
@@ -0,0 +1,16 @@
+    "category": "Warehouse Management",
+    "version": "13.0.1.0.0",
+    "license": "AGPL-3",
+    "depends": ["stock_dynamic_routing", "stock_checkout_sync"],
```
Need to add `OCA/wms` in `oca_dependencies.txt` to get `stock_dynamic_routing`?

guewen replied:
added, thanks!

sebalix commented:
The test which fails in `stock_available` is not related to this PR, I think the variants created in the failing test ( https://github.com/OCA/stock-logistics-warehouse/blob/13.0/stock_available/tests/test_stock_available.py#L31-L55 ) should have different attributes to not trigger the constraint on `flush` call.

simahawk commented:
@sebalix @guewen anything we can do to unlock this?

Commit:
Fix test in stock_available

The test creates 2 product.product on the same product.template,
which is illegal if they don't have different attribute values.

No error is raised, unless another modules runs a "flush", which
does the INSERT in database and fails with:

psycopg2.IntegrityError: duplicate key value violates unique constraint "product_product_combination_unique"
DETAIL:  Key (product_tmpl_id, combination_indices)=(31, ) already exists.

guewen commented:
I fixed the error with `stock_available`, but we have another unrelated error:

```
Traceback (most recent call last):
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/tests/test_inventory.py", line 38, in test_inventory_count_ops
operation = self._open_screen("inventory")
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/tests/common.py", line 68, in _open_screen
getattr(shuttle or self.shuttle, "switch_{}".format(mode))()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_shuttle.py", line 210, in switch_inventory
return self.action_open_screen()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_shuttle.py", line 155, in action_open_screen
operation.on_screen_open()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_base.py", line 175, in on_screen_open
self.reset_steps()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_inventory.py", line 196, in reset_steps
super().reset_steps()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_base.py", line 167, in reset_steps
self.next_step()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_base.py", line 147, in next_step
if not transition.action or transition.action(self):
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_inventory.py", line 36, in <lambda>
lambda self: self.select_next_inventory_line(),
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_inventory.py", line 242, in select_next_inventory_line
self.fetch_tray()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_inventory.py", line 231, in fetch_tray
location.fetch_vertical_lift_tray()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/stock_location.py", line 140, in fetch_vertical_lift_tray
tray.fetch_vertical_lift_tray(cell_location=self)
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/stock_location.py", line 142, in fetch_vertical_lift_tray
self._hardware_vertical_lift_fetch_tray(cell_location=cell_location)
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/stock_location.py", line 67, in _hardware_vertical_lift_fetch_tray
payload = self._hardware_vertical_lift_fetch_tray_payload(cell_location)
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/stock_location.py", line 118, in _hardware_vertical_lift_fetch_tray_payload
raise NotImplementedError()
NotImplementedError
```

This one happens because https://github.com/OCA/stock-logistics-warehouse/blob/c5aa718b4cc8ba16dfc5b8df5885d170f8587138/stock_vertical_lift/models/stock_location.py#L35-L39

is not computed.

It fails after: https://github.com/odoo/odoo/commit/3a6ac95ddf6a8719aa714a83fc339ba838c3811d

guewen commented:
Tests are now successful after rebase (fixed by #1035).

guewen commented:
Note: do not merge, will be moved to OCA/wms as asked by @jgrandguillaume

simahawk commented:
@guewen shall we close this in favor of https://github.com/OCA/wms/pull/115 ?
